inc/api/namespace.php

```diff
@@ -671,6 +671,26 @@ function get_top_data( $start, $end, ?Filter $filter = null ) {
 
 		$thumbnail = $thumbnail_id ? wp_get_attachment_image_url( $thumbnail_id, get_available_thumbnail_size() ) : '';
 
+		if ( $post->post_type === 'wp_block' ) {
+			$preview_url = sprintf(
+				get_home_url() . '?preview-block-id=%d&nonce=%s',
+				$post->ID,
+				wp_create_nonce( 'preview-block' )
```

Comment:
Could we make the nonce specific to the post ID, so it's not a shared nonce for everything?

Also we need to make sure the current user has access to view the block, do we need a cap check for that?

Reply:
Actually, I just thought... this needs a bit more consideration on how to handle private content. The issue with the nonce is it will change every 12 hours, when we don't probably want to change the URL to the `block-image` endpoint (to keep cache working.) So, I think we might need to us a hmac or something.

If we do that, we need to consider what we should do for users who don't have access to read the wp_block. I'm out for the day now but I think this logic might need to change a bit to accomadate.

Reply:
I'm thinking the trade off between caching for 12 hours only and managing access to content would probably settle for the earlier, users can be demoted/removed, and this is a way to make sure they still have access to view content.

I'm assuming whomever have access to the dashboard, currently those with `edit_posts`, will have access to all content by default.

Reply:
What I did forget to implement is adding a filter to manage disabling the feature, and maybe an additional cap check to allow for filtering private content manually.

Reply:
Added in https://github.com/humanmade/aws-analytics/pull/440/commits/bdd5d712b870d41730bf45c042f557a0171b7c64

Reply:
@shadyvb I think we need to cache for longer than 12 hours, as generating the previews is not going to be that fast.

I think fundamentally though, we need to make it so users can only view the previews if they have access to `read_post` for each post, so we should be wrapping it in checks for that.

So, I think we should:

- use a hmac for the key, instead of a nonce. A nonce is time based which I think doesn't make sense in this context. The hmac should be checked on the request to render the preview URL.
- Wrap the URL generation / image in a `current_user_can( 'read_post', $id )` to only show previews if the user has access to read that post.

Does that make sense?

Reply:
Does that solve the issue where links are saved/cached/leaked ? If I understood correctly, the checks we need will be done while generating the links not while hitting them.

Reply:
How about caching the previews in WP ? This way the preview is cached as needed, and we use WordPress safeguards to protect content. Thoughts ?

To clarify, WordPress will then request the preview URL and deliver the result to the authenticated user.